Multiple instances in sync (#1409)
OctopusDeploy/docs: docs/administration/sync-instances/index.md

:::

A rule of thumb to follow is don't have instances more than one minor version apart.  For example, the source instance runs **Octopus 2021.2**, and the destination instance runs **Octopus 2021.3**.  Ideally, all instances would be on the same `Major.Minor` version.  If you run into unexpected 400 bad request errors, the typical remediation is to upgrade both instances to the same version.

### Have a single source of truth

It is much easier to sync everything in a one-way direction.  The source instance should remain the source instance every time the syncing process runs.  It shouldn't be the source instance one day, and two days later it is the destination instance.  

When a conflict is found it will be nearly impossible to know which instance is "right" and the change should be accepted.  For example, both instances have added a step added to the same deployment process, on one instance it is a new manual intervention step, while the other instance it is a run a script step.  Should both exist?  Only one copied over?  You'd need a comparison tool similar to Beyond Compare to reconcile this conflict.

> It's okay to have known differences between the instances, such as different environments, lifecycles, variable values, tenants, deployment targets, channels, and more.  But when something new is added, such as a new variable or step, it should done on one instance and synced to the other instance.  It is hard enough to detect when something is "new".  There is no need to add to that complexity.  A one-way sync will help keep conflicts to a minimum.

Reviewer: ```suggestion
It's okay to have known differences between the instances, such as different environments, lifecycles, variable values, tenants, deployment targets, channels, and more.  But when something new is added, such as a new variable or step, it should be done on one instance and synced to the other instance.  It is hard enough to detect when something is "new". A one-way sync will help keep conflicts to a minimum, and reduce complexity.
```

The data that must exist but can have different details is:

- Infrastructure
    - Accounts: Same account type, different credentials           
    - Worker Pools: Different workers
    - Machine Policies: Different health check frequency, tentacle, and calamari update settings
- Library
    - Certificates: same certificate type, different cert
    - External Feeds: same feed type, different credentials
    - Lifecycles: different phases and retention policies
- Server Configuration
    - Teams: user role mapping, different members
- Project Groups: You don't have to sync all the projects in a project group; only the project group has to exist.

> It is possible to leverage variables instead of directly referencing that data in a deployment or runbook process.  As long as the variable exists, is of the correct type, and is assigned to something that exists on the destination instance, the deployment process will work.  For example, a variable with the name `Project.AWS.Account` is assigned to an account named `Dev/Test Account` on the source instance.  On the destination instance that same variable can be assigned to `Staging/Prod Account`. 

Reviewer: ```suggestion
It is possible to leverage variables instead of directly referencing that data in a deployment or runbook process.  The deployment process will work as long as the variable exists, is of the correct type, and is assigned to something that exists on the destination instance.  For example, a variable named `Project.AWS.Account` refers to an account called `Dev/Test Account` on the source instance.  That same variable can refer to the `Staging/Prod Account` on the destination instance. 
```

In most cases, it doesn't make much sense to sync Accounts and Certificates.  But the variables referencing the Accounts and Certificates are used in Deployment and Runbook processes.  You have a few options:

- Create an Account or Certificate with the same name on each instance but different details.  You don't have to modify any variables or processes.
- Re-use the same variable name but associate it to different Accounts or Certificates.  The syncing process will only create new variables and insert dummy data.

### Variable Scoping
Syncing variables between instances with different environments is the most complex problem.  Imagine your source instance has **Development** and **Test** while the destination instance has **Staging** and **Production**.  The source instance has the following variables.

| Variable Name             | Value        | Scope       |
| ------------------------- | ------------ | ----------- |
| Application.Database.Name | `OctoFx-Dev `  | Development |
|  | `OctoFX-Test`  | Test        |
| ConnectionString          | `Database=#{Application.Database.Name}` | |

> Syncing `ConnectionString` as-is to the destination instance makes much sense as it has no scoping.  You'll need to sync over the variable `Application.Database.Name` as `ConnectionString` references it.  But what about the values?  Those values are tied to environments that do not exist on the destination instance.  

Reviewer: ```suggestion
Syncing `ConnectionString` as-is to the destination instance makes sense as it has no scoping.  You'll need to sync over the variable `Application.Database.Name` as `ConnectionString` references it.  But what about the values?  Those values are tied to environments that do not exist on the destination instance.  
```

| Variable Name             | Value        | Scope       |
| ------------------------- | ------------ | ----------- |
| Application.Database.Name | `Replace Me`   | |
| ConnectionString          | `Database=#{Application.Database.Name}` | |

That only covers the initial sync.  Recurring syncs will add additional challenges because you'll need to account for both changed and new variables.  The fact that a variable can have multiple values, each with different scoping, makes detecting "new" variables complex.  

Variables can be scoped to:

- Environments
- Channels
- Process Owner (deployment process or runbooks)
- Deployment Process Steps
- Roles
- Deployment Targets
- Tenant Tags

Some of those items, Tenant Tags, Roles, Process Owner, will be _exactly_ alike between instances.  While other items, Environments, Channels, Deployment Process Steps, Deployment Targets, will be different.  

Imagine the variables on the source instance were changed from:

| Variable Name             | Value        | Scope       |
| ------------------------- | ------------ | ----------- |
| Application.Database.Name | `OctoFx-Dev`   | Development |

> |  | OctoFX-Test  | Test        |

Reviewer: ```suggestion
|  | `OctoFX-Test`  | Test        |
```
